File: Data Structures & Algorithms/jump-game-ii/submission-3.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)

        if n == 1:
            return 0

        e = 0
        ne = 0
        c = 0

        for i in range(n):
            ne = max(ne, i + nums[i])

            if i == e:
                e = ne
                c += 1
                if e >= n-1:
                    return c

File: Data Structures & Algorithms/sliding-window-maximum/submission-3.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> w;
        vector<int> res;
        int l = 0, r = 0;
        while (r < nums.size()) {
            while (!w.empty() and nums[w.front()] < nums[r]) {
                w.pop_front();
            }
            w.push_front(r);
            if (w.back() < l) {
                w.pop_back();
            }
            if (r >= k - 1) {
                res.push_back(nums[w.back()]);
                l++;
            }
            r++;
        }
        return res;
        
    }
};
